### DISCUSSION TITLE
Any update on Deprecation of UIWebView for IOS?

### DISCUSSION THREAD

**Discussion by choukkseyrahul** [ionic-team/capacitor#3853] [Q&A] [answered]
Hello everyone,
As Apple will start rejecting app which refer to UIWebView, so do we have any update for this?

https://developer.apple.com/news/?id=edwud51q

**Comment by jcesarmobile** [comment, marked as answer]
Capacitor doesn't use UIWebView since version 1.2.0, launched in September 2019

**Reply by choukkseyrahul** [reply]
Agreed, but I still see's UIWebView reference below:
https://github.com/ionic-team/capacitor/blob/1.x/ios/CapacitorCordova/CapacitorCordova/Classes/Public/CDVUIWebViewDelegate.m

Any update will be helpful.

**Reply by jcesarmobile** [reply]
Those references are safe to use and are there for some Cordova plugins compatibility. But have been removed in capacitor 3 alphas since most plugins are now updated.

**Reply by choukkseyrahul** [reply]
"References are safe" : Are you saying if we go with 1.4.x version deployment in IOS, then will not get rejection warning?

I can see capacitor@2.4.3 version, CDVUIWebViewDelegate have been removed as it refer UIWebView.
If I go with the latest version so there will me massive changes in several plugins which currently using in projects. Reason : many plugin have dependency of capacitor version.

**Reply by jcesarmobile** [reply]
CDVUIWebViewDelegate is not removed in 2.4.3, it’s removed in 3.0.0 (still alpha) 
https://github.com/ionic-team/capacitor/blob/2.x/ios/CapacitorCordova/CapacitorCordova/Classes/Public/CDVUIWebViewDelegate.m

You can use any version from 1.2.0 and you won’t get a rejection because of capacitor code. You can still get a rejection because of some cordova and capacitor plugin might use UIWebView, check the blog post from dotNetkow as it contains a few plugins that caused problems.

**Comment by dotNetkow** [comment]
Yep ^^^ and also, we keep this [helpful blog post](https://ionicframework.com/blog/understanding-itms-90809-uiwebview-api-deprecation/) up to date if anything changes. Recently added a section on Known Plugin Issues for some Cap/cordova plugins and how to manage them.

**Comment by choukkseyrahul** [comment]
Agreed, but I still see's UIWebView reference below:
https://github.com/ionic-team/capacitor/blob/1.x/ios/CapacitorCordova/CapacitorCordova/Classes/Public/CDVUIWebViewDelegate.m

Any update will be helpful.

**Comment by sharmakshay** [comment]
I would like to get an update on this as well.

A few things that I have gathered from reading multiple blog posts (including ones above), UIWebView has been deprecated and Apple will reject any apps that still use this. If we are using Capacitor, we need to:
1. Use capacitor version >= 1.2.0
2. Check for any references to UIWebView in cordova plugins

Below are all the references I see to UIWebView in my iOS project (all are coming from CapacitorCordova):
- **CDVCommandDelegate.h**
// Returns the User-Agent of the associated UIWebView.
- **CDVUIWebViewDelegate.h**
@interface CDVUIWebViewDelegate : NSObject <UIWebViewDelegate>{
__weak NSObject <UIWebViewDelegate>* _delegate;
(id)initWithDelegate:(NSObject <UIWebViewDelegate>*)delegate;
- **CDVUIWebViewDelegate.m**
#import "CDVUIWebViewDelegate.h"
@implementation CDVUIWebViewDelegate
(BOOL)isPageLoaded:(UIWebView*)webView
(BOOL)isJsLoadTokenSet:(UIWebView*)webView
 ... (and many more)
- **CDVUserAgentUtil.m**
// TODO: We should figure out how to update the user-agent of existing UIWebViews when this happens.
// Setting the UserAgent must occur before a UIWebView is instantiated.
- **CapacitorCordova-Umbrella.h**
#import "CDVUIWebViewDelegate.h"

From my understanding these should not trigger Apple's check (as mentioned in the blog link above)
> You may notice “UIWebView” references within your Capacitor-based iOS project. Files within the “CapacitorCordova” pod, for example, relate to Cordova compatibility. These will not trigger the warning.

So, not sure why we are still unable to deploy our app onto App Store Connect. We see the following error:
_ITMS-90809: Deprecated API Usage - New apps that use UIWebView are no longer accepted. Instead, use WKWebView for improved security and reliability. Learn more (https://developer.apple.com/documentation/uikit/uiwebview)._

Any help would be greatly appreciated.

**Reply by sharmakshay** [reply]
Just to follow up on the above in case anyone else is stuck on the same issue; I was able to resolve this by doing a search for _UIWebView_ in the ionic project instead of just iOS directory/project. Here I found that my ionic and firebase dependencies (such as ionic/react, ionic/cli, ionic/react-router, firebase) were out of date since they were still referencing _UIWebView_. Once I updated these dependencies, I was able to successfully deploy my app onto App Store Connect.
